sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -153,6 +155,336 @@ object AggregateModeInfo {
+        preStepAttributes.toList)))
+    } else {
+      (Some(GpuBindReferences.bindGpuReferences(preStep, preStepAttributes.toList)), None)
+    }
+
+    // a bound expression that is applied after the cuDF aggregate
+    private val postStepBound =
+      GpuBindReferences.bindGpuReferences(postStep, postStepAttr)
+
+    /**
+     * Apply the "pre" step: preMerge for merge, or pass-through in the update case
+     * @param toAggregateBatch - input (to the agg) batch from the child directly in the
+     *                         merge case, or from the `inputProjection` in the update case.
+     * @return a pre-processed batch that can be later cuDF aggregated
+     */
+    def preProcess(
+        toAggregateBatch: ColumnarBatch,
+        metrics: GpuHashAggregateMetrics): SpillableColumnarBatch = {
+      val projectedCb = withResource(toAggregateBatch) { inputBatch  =>
+        withResource(new NvtxRange("pre-process", NvtxColor.DARK_GREEN)) { _ =>
+          if (useTieredProject) {
+            preStepBoundTiered.get.tieredProject(inputBatch)
+          } else {
+            GpuProjectExec.project(inputBatch, preStepBound.get)
+          }
+        }
+      }
+      SpillableColumnarBatch(
+        projectedCb,
+        SpillPriorities.ACTIVE_BATCHING_PRIORITY,
+        metrics.spillCallback)
+    }
+
+    def aggregate(preProcessed: ColumnarBatch): ColumnarBatch = {
+      if (groupingOrdinals.nonEmpty) {
+        performGroupByAggregation(preProcessed)
+      } else {
+        performReduction(preProcessed)
+      }
+    }
+
+    def aggregate(
+        metrics: GpuHashAggregateMetrics,
+        preProcessed: SpillableColumnarBatch): SpillableColumnarBatch = {
+      val aggregatedSeq =
+        withRetry(preProcessed, splitSpillableInHalfByRows) { preProcessedAttempt =>
+          withResource(preProcessedAttempt.getColumnarBatch()) { cb =>
+            SpillableColumnarBatch(
+              aggregate(cb),
+              SpillPriorities.ACTIVE_BATCHING_PRIORITY,
+              preProcessed.getSpillCallback)
+          }
+        }.toSeq
+
+      // We need to merge the aggregated batches into 1 before calling post process,
+      // if the aggregate code had to split on a retry
+      if (aggregatedSeq.size > 1) {
+        val concatted = concatenateBatches(metrics, aggregatedSeq)
+        withRetryNoSplit(concatted) { attempt =>
+          withResource(attempt.getColumnarBatch()) { cb =>
+            SpillableColumnarBatch(
+              aggregate(cb),
+              SpillPriorities.ACTIVE_BATCHING_PRIORITY,
+              concatted.getSpillCallback)
+          }
+        }
+      } else {
+        aggregatedSeq.head
+      }
+    }
+
+    /**
+     * Invoke reduction functions as defined in each `CudfAggreagte`
+     * @param preProcessed - a batch after the "pre" step
+     * @return
+     */
+    def performReduction(preProcessed: ColumnarBatch): ColumnarBatch = {
+      withResource(new NvtxRange("reduce", NvtxColor.BLUE)) { _ =>
```

Comment:
Do we want/need to allow this to have a withRetry? I know it is less likely to happen because the reduction should be smaller in terms of memory, but it might be good to at least have a follow on for this.

Reply:
It is under a `withRetry` already. We wrap a higher level method `aggregate` with the retry block, and it can be either a reduction or a group by.